File: monay-wallet-legacy/Monay_Wallet_Backend/src/models/Transaction.js
module.exports = (sequelize, DataTypes) => {
    const Transaction = sequelize.define(
        'Transaction',
        {
            fromUserId: {
                type: DataTypes.INTEGER
            },
            toUserId: {
                type: DataTypes.INTEGER
            },
            transactionUserId: {
                type: DataTypes.INTEGER
            },
            transactionId: {
                type: DataTypes.STRING(255)
            },
            amount: {
                type: DataTypes.FLOAT
            },
            message: {
                type: DataTypes.TEXT
            },
            transactionType: {
                type: DataTypes.ENUM('credit', 'debit')
            },
            paymentMethod: {
                type: DataTypes.ENUM('card', 'wallet', 'account')
            },
            actionType: {
                type: DataTypes.ENUM('deposit', 'withdrawal', 'transfer')
            },
            paymentStatus: {
                type: DataTypes.ENUM('pending', 'completed', 'cancelled', 'failed'),
                defaultValue: 'completed'
            },
            status: {
                type: DataTypes.ENUM('failed', 'success')
            },
            cardType: {
                type: DataTypes.STRING(255)
            },
            cardName: {
                type: DataTypes.STRING(255)
            },
            bankName: {
                type: DataTypes.STRING(255)
            },
            last4Digit: {
                type: DataTypes.STRING(10)
            },
            apiReponse: {
                type: DataTypes.TEXT
            },
            cnpToken: {
                type: DataTypes.STRING(255)
            },
            paymentRequestId: {
                type: DataTypes.INTEGER
            },
            parentId: {
                type: DataTypes.INTEGER
            }
        },
        {
            underscored: true
        },
    );
    Transaction.associate = function (models) {
        Transaction.belongsTo(models.User, {
            foreignKey: 'fromUserId', onDelete: 'cascade', as: 'fromUser'
        });
        Transaction.belongsTo(models.User, {
            foreignKey: 'toUserId', onDelete: 'cascade', as: 'toUser'
        });
        Transaction.belongsTo(models.User, {
            foreignKey: 'transactionUserId', onDelete: 'cascade', as: 'transactionUser'
        });
        Transaction.belongsTo(models.PaymentRequest, {
            foreignKey: 'paymentRequestId', onDelete: 'set null', onUpdate: 'set null'
        });
    };
    return Transaction;
};

File: monay-wallet/ios/monay/Monay/Model/Common/Transaction.swift
//
//  Transaction.swift
//  Monay
//
//  Created by WFH on 22/08/20.
//  Copyright © 2020 Codiant. All rights reserved.
//

import UIKit
import ObjectMapper

class Transaction: Mappable {
    
    var id: String? 
    var fromUserId: String?
    var toUserId: String?
    var amount: String?
    var actionType: ActionType?
    var createdAt: String?
    var fromUser: User?
    var toUser: User?
    var status: String?
    var transactionId: String?
    var paymentStatus: PaymentStatus?
    var message: String?
    var cardType: String?
    var cardName: String?
    var last4Digit: String?
    var cardIconUrl: String?
    var bankName: String?
    var paymentMethod: String?
    
    required init?(map: Map) {}
    
    func mapping(map: Map) {
        id <- (map["id"], intToStringTransform)
        fromUserId <- (map["fromUserId"], intToStringTransform)
        toUserId <- (map["toUserId"], intToStringTransform)
        amount <- (map["amount"], intToStringTransform)
        
        if amount == nil {
            amount <- (map["amount"], doubleToStringTransform)
        }
        
        actionType <- (map["actionType"], EnumTransform<ActionType>())
        createdAt <- map["createdAt"]
        fromUser <- map["fromUser"]
        toUser <- map["toUser"]
        status <- map["status"]
        transactionId <- map["transactionId"]
        paymentStatus <- (map["paymentStatus"], EnumTransform<PaymentStatus>())
        message <- map["message"]
        cardType <- map["cardType"]
        cardName <- map["cardName"]
        last4Digit <- map["last4Digit"]
        cardIconUrl <- map["cardIconUrl"]
        bankName <- map["bankName"]
        paymentMethod <- map["paymentMethod"]
    }
}
